File: Programming languages/parciales/parcial03/cajacambios/cambios.rb

class CajaDeCambios
	def initialize
		@estado = :neutra
		@embrague = :libre
		@cambio_puesto = false
	end
	def establecerEmb
	  if @embrague == :libre
    	@embrague = :montado
      end
	end
	def liberarEmb
	  if @cambioPuesto
   		@embrague == :libre
        @cambioPuesto = false
   	  end
	end
		def ponerNeutra
		if @embrague == :montado 
    	@estado = :neutra
		end
	end

	def ponerPrimera
		if @embrague == :montado 
    	@estado = :primera
		end
	end
	def ponerSegunda
		if @embrague == :montado
			@estado = :segunda
		end
	end
	def ponerTercera
		if @embrague == :montado
			@estado = :tercera
		end
	end
	def ponerCuarta
		if @embrague == :montado
			@estado = :cuarta
		end
	end
	def ponerQuinta
		if @embrague == :montado
			@estado = :quinta
		end
	end
	def ponerReversa
		if @embrague == :montado
			@estado = :reversa
		end
	end
	def obtenerCambio
		@estado
	end
end

File: Programming languages/parciales/parcial02/cajacambios/cambios.cpp
#include "cambios.h"

CajaDeCambios::CajaDeCambios(): estado(neutra), embrague(libre), cambioPuesto(false) {}

CajaDeCambios::~CajaDeCambios() {}

void CajaDeCambios::establecerEmbr() {
  if (embrague == libre) {
    embrague = montado;
  }
}

void CajaDeCambios::liberarEmb() {
  if (cambioPuesto) {
    embrague = libre;
    cambioPuesto = false;
  }
}

void CajaDeCambios::ponerPrimera() {
  if (embrague == montado) {
    estado = primera;
  }
}

void CajaDeCambios::ponerSegunda() {
  if (embrague == montado) {
    estado = segunda;
  }
}

void CajaDeCambios::ponerTercera() {
  if (embrague == montado) {
    estado = tercera;
  }
}

void CajaDeCambios::ponerCuarta() {
  if (embrague == montado) {
    estado = cuarta;
  }
}

void CajaDeCambios::ponerQuinta() {
  if (embrague == montado) {
    estado = quinta;
  }
}

void CajaDeCambios::ponerReversa() {
  if (embrague == montado) {
    estado = reversa;
  }
}

void CajaDeCambios::ponerNeutra() {
  if (embrague == montado) {
    estado = neutra;
  }
}
EstadoCambio CajaDeCambios::obtenerCambio() const {
  return estado;
}